### PR TITLE
[SPARK-49565][SQL] Add SQL pipe syntax for the FROM operator

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR adds SQL pipe syntax support for the FROM operator.

Note: this PR includes all content in https://github.com/apache/spark/pull/48724 and also fixes remaining tests.

For example:

```
CREATE TABLE t(x INT, y STRING) USING CSV;
INSERT INTO t VALUES (0, 'abc'), (1, 'def');

CREATE TABLE other(a INT, b INT) USING JSON;
INSERT INTO other VALUES (1, 1), (1, 2), (2, 4);

FROM t
|> SELECT 1 AS X;
```

### Why are the changes needed?

This allows users to use FROM in the pipe syntax format.

### Does this PR introduce any user-facing change?

Yes, as indicated in the example.

### How was this patch tested?

* Unit tests in `SparkSqlParserSuite` and `PlanParserSuite`.
* Golden file based end to end tests in `pipe-operators.sql`.

Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @cloud-fan @dongjoon-hyun here is the PR with test cases fixed.

**Review by dongjoon-hyun — COMMENTED** [review]
Could you fix the UT failure?
```
[info] *** 1 TEST FAILED ***
[error] Failed: Total 7714, Failed 1, Errors 0, Passed 7713, Ignored 5
[error] Failed tests:
[error] 	org.apache.spark.sql.catalyst.parser.PlanParserSuite
```

**Comment by dtenedor** [comment]
Thanks @dongjoon-hyun, fixed the unit test and manually re-ran it again to make sure.

image

File: sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4 [hunk 1]
```diff
@@ -648,7 +648,7 @@ sortItem
     ;
 
 fromStatement
-    : fromClause fromStatementBody+
+    : fromClause (fromStatementBody)*
```
Comment (cloud-fan) [inline sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4:651]: super nit: is it the same as `fromStatementBody*`?

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you, @deanchen and @cloud-fan .

But, the authorship looks incorrect to me.

As I gave you the direction, https://github.com/apache/spark/pull/48724#issuecomment-2521124627, when we start to take over a PR, we usually keep the original author's code in the first commit to give a credit to the original author. I don't find the original author's commit here, @dtenedor .

Screenshot 2024-12-10 at 13 45 45

**Comment by dongjoon-hyun** [comment]
cc @jiashenC as the original author.

**Comment by dongjoon-hyun** [comment]
At least, please add an empty commit with @jiashenC 's authorship via the following CLI command.
```
git commit --allow-empty --author=....
```

**Comment by dtenedor** [comment]
@dongjoon-hyun sounds good, I added the empty commit as suggested. In the future I can include the `--author=...` option as well.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for the update, @dtenedor .

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0 on February 2025.

Thank you all!

**Comment by dongjoon-hyun** [comment]
For the record, this is merged with the following authorships.
```
Lead-authored-by: Jiashen Cao <jiashenc@gatech.edu>
Co-authored-by: Daniel Tenedorio <daniel.tenedorio@databricks.com>
```

**Comment by dongjoon-hyun** [comment]
To @jiashenC , do you have an Apache JIRA ID? Please let me know your ID.

**Comment by dtenedor** [comment]
@jiashenC it would be awesome to collaborate on Spark with you more in the future. If you're interested.

**Comment by jiashenC** [comment]
@dongjoon-hyun Sorry for slow response, I believe my JIRA ID is just `jiashenc`. 

@dtenedor Thank you for covering this for me! Yeah, I am definitely interested in contributing to Spark.

**Comment by dongjoon-hyun** [comment]
I added `jiashenc` to the Apache Spark contributor group and assigned SPARK-49565 to you, @jiashenC .

Welcome to the Apache Spark community and congratulations for your first commit!

**Pull request closed without merging**